File: core/FindTheClosestPalindrome/FindTheClosestPalindrome.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class FindTheClosestPalindrome {
    public String nearestPalindromic(String n) {
        int length = n.length();
        long int_n = Long.parseLong(n);
        long prefix = Long.parseLong(n.substring(0, (length + 1) / 2));
        List<Long> candidate = new ArrayList<>();
        candidate.add((long) (Math.pow(10, (length - 1)) - 1));
        candidate.add((long) (Math.pow(10, length) + 1));
        for (long i = prefix - 1; i < prefix + 2; i++) {
            long y = length % 2 == 0 ? i : i / 10;
            long x = i;
            while (y != 0) {
                x = x * 10 + y % 10;
                y /= 10;
            }
            candidate.add(x);
        }

        long res = -1;
        for (long candi : candidate) {
            if (candi - int_n != 0) {
                if (res == -1 || Math.abs(int_n - candi) < Math.abs(int_n - res)) {
                    res = candi;
                }
            }
        }
        return Long.toString(res);
    }
}

File: core/FindTheClosestPalindrome/FindTheClosestPalindrome.py
def nearestPalindromic(n: str) -> str:
    left, right, temp, length = 0, len(n) - 1, list(n), len(n)
    int_n = int(n)
    candidate = [10 ** (length - 1) - 1, 10 ** length + 1]
    self_prefix = int(n[:(length + 1) // 2])
    for prefix in range(self_prefix - 1, self_prefix + 2):
        y = prefix if length % 2 == 0 else prefix // 10
        while y != 0:
            prefix = prefix * 10 + y % 10
            y //= 10
        candidate.append(prefix)

    res = -1
    for candi in candidate:
        if str(candi) != n:
            if res == -1 or abs(int_n - candi) < abs(int_n - res):
                res = candi

    return str(res)


if __name__ == '__main__':
    print(nearestPalindromic("1283"))
